### PR TITLE
Print partitions summary

### PR DESCRIPTION
**Pull request by petrohi**
- Support printing partitions summary;
- `tensil compile` command line flags for summaries;
- Auto-detect K/M/G/T letter and divisor;
- Print totals in instructions summaries;
- Print MAC efficiency in scheduler summary;
- Reference graph node names in layer name.

### CODE REVIEW

File: tools/src/tensil/tools/compiler/BackendStream.scala [hunk 1]
```diff
@@ -50,9 +77,9 @@ object BackendStats {
       )
   }
 
-  def printCompositionSummary(stats: BackendStats) {
+  def printCompositionSummary(name: String, stats: BackendStats) {
```
Comment (tdb-alcorn) [inline tools/src/tensil/tools/compiler/BackendStream.scala:80]: what is name intended to be here? model name?

File: tools/src/tensil/tools/compiler/Scheduler.scala [hunk 2]
```diff
@@ -740,14 +773,25 @@ class Scheduler(
...
+        tb.addNamedLine("MAC efficiency (%)", macEfficiency * 100f)
       print(tb)
+    }
 
-      if (collectStats) {
-        BackendStats.printCompositionSummary(layerStats.get)
-        BackendStats.printCyclesSummary(layerStats.get)
-        BackendStats.printEnergySummary(layerStats.get)
+    if (collectStats) {
+      if (options.printInstructionsSummary) {
```
Comment (tdb-alcorn) [inline tools/src/tensil/tools/compiler/Scheduler.scala:788]: It feels a bit odd to call this the "instructions summary" when it also gates the metrics. Probably fine for now but if you can think of a better name might be worth changing.

**Review by tdb-alcorn — APPROVED** [review]
Lgtm, two quick thoughts

File: tools/src/tensil/tools/compiler/Scheduler.scala (hunk 2 above)
Comment (petrohi) [reply tools/src/tensil/tools/compiler/Scheduler.scala:788]: Happy to change to something better. "Instructions summary" prints the following tables.

```
-----------------------------
TOTAL INSTRUCTIONS COMPOSITION (NUMBER/TOTAL SIZE)
-----------------------------
MatMul:                 2  2
DataMove(DRAM0ToLocal): 2  2
DataMove(DRAM1ToLocal): 2  34
DataMove(AccToLocal):   2  2
SIMD:                   2  2
DataMove(LocalToDRAM0): 2  2
LoadWeights:            2  34
TOTAL:                  14 78
-----------------------------
--------------------------------------
TOTAL INSTRUCTIONS LATENCY
--------------------------------------
MatMul:                 2.000   Cycles
DataMove(DRAM0ToLocal): 8.000   Cycles
DataMove(DRAM1ToLocal): 136.000 Cycles
DataMove(AccToLocal):   2.000   Cycles
SIMD:                   2.000   Cycles
DataMove(LocalToDRAM0): 4.000   Cycles
LoadWeights:            34.000  Cycles
TOTAL:                  188.000 Cycles
--------------------------------------
--------------------------------------
TOTAL INSTRUCTIONS ENERGY
--------------------------------------
MatMul:                 512.000 Units
DataMove(DRAM0ToLocal): 400.000 Units
DataMove(DRAM1ToLocal): 6.800   KUnits
DataMove(AccToLocal):   20.000  Units
SIMD:                   32.000  Units
DataMove(LocalToDRAM0): 400.000 Units
LoadWeights:            340.000 Units
TOTAL:                  8.504   KUnits
--------------------------------------
```

What will describe these better?

Comment (tdb-alcorn) [reply tools/src/tensil/tools/compiler/Scheduler.scala:788]: Ah, my mistake, I somehow missed that this was the actual breakdown per instruction and not the overall metrics. Name is fine as it is!

**Pull request merged**